File: solutions/elixir/allergies/3/lib/allergies.ex
defmodule Allergies do
  @allergies [
    {"cats", 128},
    {"pollen", 64},
    {"chocolate", 32},
    {"tomatoes", 16},
    {"strawberries", 8},
    {"shellfish", 4},
    {"peanuts", 2},
    {"eggs", 1}
  ]

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags), do: flags |> list_allergies()

  @doc false
  defp list_allergies(flags, allergies \\ @allergies, list \\ [])
  defp list_allergies(_, [], list), do: list
  defp list_allergies(flags, allergies, list) when flags > 255 do
    list_allergies(flags - 256, allergies, list)
  end
  defp list_allergies(flags, [{allergy, bit} | tail], list) do
    cond do
      bit <= flags -> list_allergies(flags - bit, tail, [allergy | list])
      bit > flags -> list_allergies(flags, tail, list)
    end
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    flags
    |> list()
    |> Enum.member?(item)
  end
end

File: solutions/ocaml/allergies/2/allergies.ml
open Base

type allergen =
  | Eggs
  | Peanuts
  | Shellfish
  | Strawberries
  | Tomatoes
  | Chocolate
  | Pollen
  | Cats

let allergen_scores =
  [
    (Cats, 128);
    (Pollen, 64);
    (Chocolate, 32);
    (Tomatoes, 16);
    (Strawberries, 8);
    (Shellfish, 4);
    (Peanuts, 2);
    (Eggs, 1);
  ]

let allergies score =
  let score = if score > 256 then score % 256 else score in
  let _, list =
    List.fold_left allergen_scores ~init:(score, [])
      ~f:(fun (score, list) (a, s) ->
        if score >= s then (score - s, a :: list) else (score, list))
  in
  list

let allergic_to score allergen =
  List.mem ~equal:phys_equal (allergies score) allergen

(* let allergens = [ Eggs; Peanuts; Shellfish; Strawberries; Tomatoes; Chocolate; Pollen; Cats ] *)

(* let score = function *)
  (* | Eggs -> 1 *)
  (* | Peanuts -> 2 *)
  (* | Shellfish -> 4 *)
  (* | Strawberries -> 8 *)
  (* | Tomatoes -> 16 *)
  (* | Chocolate -> 32 *)
  (* | Pollen -> 64 *)
  (* | Cats -> 128 *)

(* let allergic_to flag allergen = flag land score allergen = score allergen *)
(* let allergies flag = List.filter allergens ~f:(allergic_to flag) *)
